lib/crypto-algorithms/megolm.js

```diff
@@ -80,28 +80,96 @@ MegolmEncryption.prototype._ensureOutboundSession = function() {
         key.key, key.chain_index
     );
 
+    // send the keys to each (unblocked) device in the room.
+    var payload = {
+        type: "m.room_key",
+        content: {
+            algorithm: olmlib.MEGOLM_ALGORITHM,
+            room_id: this._roomId,
+            session_id: session_id,
+            session_key: key.key,
+            chain_index: key.chain_index,
+        }
+    };
+
+    var roomMembers = utils.map(room.getJoinedMembers(), function(u) {
+        return u.userId;
+    });
+
     var self = this;
-    // TODO: initiate key-sharing
-    this._prepPromise = q.delay(3000).then(function() {
-        console.log("woop woop, we totally shared the keys");
+    var txnBase = '' + (new Date().getTime()) + '.';
+    var txnCtr = 0;
+
+    // TODO: we need to give the user a chance to block any devices or users
+    // before we send them the keys; it's too late to download them here.
+    this._prepPromise = this._crypto.downloadKeys(
+        roomMembers, false
+    ).then(function(res) {
+        return self._crypto.ensureOlmSessionsForUsers(roomMembers);
+    }).then(function(devicemap) {
+        var promises = [];
+        for (var userId in devicemap) {
+            if (!devicemap.hasOwnProperty(userId)) {
+                continue;
+            }
+
+            var devices = devicemap[userId];
+
+            for (var deviceId in devices) {
+                if (!devices.hasOwnProperty(deviceId)) {
+                    continue;
+                }
+
+                var deviceInfo = devices[deviceId].device;
+                var encryptedContent = olmlib.encryptMessageForDevices(
+                    self._deviceId,
+                    self._olmDevice,
+                    [deviceInfo.getIdentityKey()],
```

Comment:
Yeah, we will, hence the need for retries.

My main concern was around how many copies of the key we can fit in a single event; it's certainly not valid to assume that one will be sufficient. And we'll need to update it anyway to use OOB events.

So the question really becomes: how should this work when OOB events land? Should we be sending an OOB event per device, or grouping them together in some way? We could maybe group them by user, but it's not even valid to assume that all the messages for a single user will fit in a single event.